src/AgentCommerce/Subscriber/WebhookSubscriber.php

```
@@ -66,22 +74,49 @@ public function handleWebhookLifecycle(EntityWrittenEvent $event): void
 
         $criteria = new Criteria(array_keys($mapped));
         $criteria->addFilter(new EqualsFilter('typeId', SwagPayPal::SALES_CHANNEL_TYPE_AGENT_COMMERCE));
-        $criteria->addAssociations([
-            'country',
-            'countries',
-            'currency',
-            'domains',
-            'productExports.storefrontSalesChannel.domains',
-            'productExports.storefrontSalesChannel.hreflangDefaultDomain',
-        ]);
 
-        $result = $this->salesChannelRepository->search($criteria, $event->getContext());
-        foreach ($result as $salesChannel) {
-            if ($mapped[$salesChannel->getId()]) {
-                $this->webhook->register($salesChannel, $event->getContext());
-            } else {
-                $this->webhook->unregister($salesChannel, $event->getContext());
+        $userId = null;
+        $context = $event->getContext();
+        $source = $context->getSource();
+        if ($source instanceof AdminApiSource) {
+            $userId = $source->getUserId();
+        }
+
+        /** @var list<string> $salesChannelIds */
+        $salesChannelIds = $this->salesChannelRepository->searchIds($criteria, $context)->getIds();
+        foreach ($salesChannelIds as $salesChannelId) {
+            try {
+                if ($mapped[$salesChannelId]) {
+                    $response = $this->webhookService->register($salesChannelId, $context);
+                } else {
+                    $response = $this->webhookService->deregister($salesChannelId, $context);
+                }
+
+                if (!$userId) {
+                    continue;
+                }
+
+                $content = json_decode($response->getBody()->getContents(), true);
```

So you're defining success or failure based on the response by decoding it _again_. I would propose to throw an exception for a request failure too, with `$content['message']` as the exception message

---

Don't think we should throw an exception and interrupt the general flow.
We are logging, we are creating a notification. Isn't this enough?

---

I see - what bothers me as that you have to make decisions outside the service if the result is to your liking, being very verbose by returning the response and decoding the content twice to access the same data - this is not ideal 🤔 
I agree that an exception is not a good idea, how about returning a result struct?

---

Referring to this: https://github.com/shopware/SwagPayPal/pull/367#discussion_r2368523312 (FYI @mstegmeyer)

A struct would be better than an array, but as the actual response?
That should now be decided before I revise it again.

---

For the Controller calling this class, I think, the response would be more sensible